src/libs/elektra/changetracking.c

```diff
@@ -0,0 +1,601 @@
+ * @param old old key
+ * @return true if @p new and @p old have a different value
+ */
+static bool keyValueDifferent (Key * new, Key * old)
+{
+	if (new->keyData == old->keyData)
+	{
+		return false;
+	}
+
+	if (new->keyData == NULL || old->keyData == NULL)
+	{
+		// one is NULL, the other one isn't -> modified
+		return true;
+	}
+
+	if (keyIsString (new) != keyIsString (old))
+	{
+		// different data types -> modified
+		return true;
+	}
+
+	if (keyIsString (new))
+	{
+		if (strcmp (keyString (new), keyString (old)) != 0)
+		{
+			// different value -> modified
+			return true;
+		}
+	}
+	else
+	{
+		if (new->keyData->dataSize != old->keyData->dataSize)
+		{
+			// different datasize -> modified
+			return true;
+		}
+
+		if (memcmp (new->keyData->data.v, old->keyData->data.v, new->keyData->dataSize) != 0)
+		{
+			// different value -> modified
+			return true;
+		}
+	}
+
+	return false;
+}
+
+/**
+ * Find the differences between two keysets
+ *
+ * @param new the new keyset
+ * @param old the old keyset
+ * @param addedKeys adds keys present in @p new but not in @p old
+ * @param removedKeys adds keys present in @p old but not in @p new
+ * @param modifiedKeys adds keys present in both @p new and @p old, but with changes in value or in the meta keys
+ * @param parentKey parent key - if this parameter is not NULL, only keys below or same are processed.
+ */
+static void findDifferences (KeySet * new, KeySet * old, KeySet * addedKeys, KeySet * removedKeys, KeySet * modifiedKeys,
+			     const Key * parentKey)
+{
+	KeySet * metaAdded = ksNew (0, KS_END);
+	KeySet * metaRemoved = ksNew (0, KS_END);
+	KeySet * metaModified = ksNew (0, KS_END);
+
+	for (elektraCursor itOld = 0; itOld < ksGetSize (old); itOld++)
+	{
+		Key * needle = ksAtCursor (old, itOld);
+
+		if (parentKey != NULL && !keyIsBelowOrSame (parentKey, needle))
+		{
+			continue;
+		}
+
+		Key * found = ksLookup (new, needle, 0);
+
+		if (found == NULL)
+		{
+			// key is present in old key set, but not in new --> removed
+			ksAppendKey (removedKeys, needle);
+		}
+		else
+		{
+			// key is present in both new and old key sets --> check for modifications
+
+			if (keyValueDifferent (found, needle))
+			{
+				// The value of the key changed --> modified
+				ksAppendKey (modifiedKeys, needle);
+			}
+			else if (keyGetNamespace (found) != KEY_NS_META)
+			{
+				// Check whether something in the meta keys has changed
+
+				KeySet * oldMeta = keyMeta (needle);
+				KeySet * newMeta = keyMeta (found);
+
+				ksClear (metaAdded);
+				ksClear (metaRemoved);
+				ksClear (metaModified);
+
+				findDifferences (newMeta, oldMeta, metaAdded, metaRemoved, metaModified, NULL);
+
+				if (ksGetSize (addedKeys) > 0 || ksGetSize (removedKeys) > 0 || ksGetSize (modifiedKeys) > 0)
+				{
+					// there was a change in the meta keys --> modified
+					ksAppendKey (modifiedKeys, needle);
+				}
+			}
+		}
+	}
+
+	for (elektraCursor itNew = 0; itNew < ksGetSize (new); itNew++)
```

Comment:
If you do the simultaneous iteration and `keyCmp` from above, this loop is also unnecessary.